src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.buffer
+import kotlinx.coroutines.flow.callbackFlow
+import kotlinx.coroutines.suspendCancellableCoroutine
+
+/** Controller to interact with the CHIP device. */
+class MatterControllerImpl(params: ControllerParams) :
+  MatterController, chip.devicecontroller.ChipDeviceController.CompletionListener {
+  private val deviceController: ChipDeviceController
+  private var completionListener: MatterController.CompletionListener? = null
+  private var nodeId: Long? = null
+
+  override fun setCompletionListener(listener: MatterController.CompletionListener) {
+    completionListener = listener
+  }
+
+  override fun pairDevice(
+    nodeId: Long,
+    address: String,
+    port: Int,
+    discriminator: Int,
+    pinCode: Long,
+  ) {
+    this.nodeId = nodeId
+    deviceController.setCompletionListener(this)
+    deviceController.pairDeviceWithAddress(nodeId, address, port, discriminator, pinCode, null)
+  }
+
+  override fun unpairDevice(nodeId: Long) {
+    deviceController.unpairDevice(nodeId)
+    this.nodeId = null
+  }
+
+  override fun establishPaseConnection(
+    nodeId: Long,
+    address: String,
+    port: Int,
+    setupPincode: Long
+  ) {
+    deviceController.establishPaseConnection(nodeId, address, port, setupPincode)
+  }
+
+  override fun onConnectDeviceComplete() {
+    completionListener?.onConnectDeviceComplete()
+  }
+
+  override fun onStatusUpdate(status: Int) {
+    completionListener?.onStatusUpdate(status)
+  }
+
+  override fun onPairingComplete(errorCode: Int) {
+    completionListener?.onPairingComplete(errorCode)
+  }
+
+  override fun onPairingDeleted(errorCode: Int) {
+    completionListener?.onPairingDeleted(errorCode)
+  }
+
+  override fun onNotifyChipConnectionClosed() {
+    completionListener?.onNotifyChipConnectionClosed()
+  }
+
+  override fun onCloseBleComplete() {
+    logger.log(Level.INFO, "Not implemented, override the abstract function.")
+  }
+
+  override fun onCommissioningComplete(nodeId: Long, errorCode: Int) {
+    completionListener?.onCommissioningComplete(nodeId, errorCode)
+  }
+
+  override fun onCommissioningStatusUpdate(nodeId: Long, stage: String?, errorCode: Int) {
+    completionListener?.onCommissioningStatusUpdate(nodeId, stage, errorCode)
+  }
+
+  override fun onReadCommissioningInfo(
+    vendorId: Int,
+    productId: Int,
+    wifiEndpointId: Int,
+    threadEndpointId: Int
+  ) {
+    completionListener?.onReadCommissioningInfo(
+      vendorId,
+      productId,
+      wifiEndpointId,
+      threadEndpointId
+    )
+  }
+
+  override fun onOpCSRGenerationComplete(csr: ByteArray) {
+    completionListener?.onOpCSRGenerationComplete(csr)
+  }
+
+  override fun onError(error: Throwable) {
+    completionListener?.onError(error)
+  }
+
+  override fun subscribe(request: SubscribeRequest): Flow<SubscriptionState> {
+    // To prevent potential issues related to concurrent modification, assign
```

Comment:
Could we explain the use case here? Since we are using flows/coroutines, I wonder if thread safety is actually an issue.

Is the issue that callbackFlow may not execute right away?

Reply:
While coroutines and flows do provide mechanisms for concurrency management, it's still important to ensure that you don't access mutable properties in a coroutine context if there is a possibility of them being modified concurrently from other threads or coroutines.

That is why I capture the value of nodeId in a local variable, this eliminate the possibility of it changing between the check and its actual use. 

callbackFlow not executing right away. Coroutines and flows execute asynchronously, and the execution of the code within a callbackFlow block will occur when it is collected by a downstream consumer.